### ISSUE TITLE
Error decoding environment variables with passenger 5.x

### ISSUE THREAD

**Issue by tomhughes** [phusion/passenger#1447]
A couple of times with passenger 5 I have seen it get stuck failing to start any application processes with this error:

```
App 9940 stderr: [ 2015-03-21 06:53:09.7076 10013/0x000000023cea90(Worker 1) utils.rb:85 ]: *** Exception RuntimeError in Rack application object (Missing `secret_token` and `secret_key_base` for 'production' environment, set these values in `config/secrets.yml`) (process 10013, thread 0x000000023cea90(Worker 1)):
```

The secret in question is actually supposed to be provided by a `SetEnv` in the apache config which is then expanded in `config/secrets.yml`.

At the point when this starts the apache log contains this lline:

```
[ 2015-03-21 06:52:58.4315 9783/7ffb7831a700 App/Spawner.h:1149 ]: Unable to decode base64-encoded environment variables: Uxxxu^@^@^@^@^@^@^@^@^QÝ^A^@^@^@^@^@
```

Note that I've elided most of the base 64 as it contains our secret, but I have confirmed that it was valid base 64 up to the trailing control characters.

So it looks like some junk padding is sometimes added to the end of the variables which then fails to decode.

**Comment by tomhughes** [comment]
Looking more closely it appears that although the section up to the control characters is valid base 64 it is not complete as it cuts off part way through the value of a variable so it appears the encoded variables are being corrupted somewhere along the line.

**Comment by tomhughes** [comment]
This happened again when the logs rolled over this morning. This time there was apparent corruption at the start and the end of the string:

```
[ 2015-03-24 06:47:45.0766 18245/7fc98c6b3700 App/Spawner.h:1149 ]: Unable to decode base64-encoded environment variables: ^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@YXxxxEs^@^@^@^@^@^@^@^@^@^@^@^@^@
```

**Comment by OnixGH** [comment]
Thanks for all the info, we'll need a bit of time to analyze this.

**Comment by FooBarWidget** [comment]
@tomhughes Is your secret a plain ASCII string? Or is it UTF-8? Or does it even contain binary data?

We haven't been able to reproduce this issue. Are you able to reliably reproduce this issue? What about in development?

**Comment by tomhughes** [comment]
It's an ASCII string. In fact it's a base 64 string - basically the result of base64 encoding some random bytes.

I can't reproduce it on demand, but if I just leave things running then it seems to happen every few days when the apache logs roll over and apache is reloaded.

Given the nature of the bug my guess would be that's not necessarily anything to do with the code that handles the encoding and decoding but could just be a bad memory access somewhere else that just happens to hit this buffer. Of course that makes finding it a major pain... Certainly when I looked at the code handling the encoding and decoding I couldn't see any obvious issues.

Is that buffer long lived? When the reload happens will it be reusing the buffer from the previous start which may have been corrupted at any point in the last 24 hours or will it generate a new one giving a much shorter window where the corruption can occur?

**Comment by FooBarWidget** [comment]
The envvarsCache buffer is supposed to live until an Apache reload happens. I've tested what happens when I perform a reload, and the buffer is invalidated as expected. I haven't encountered any memory corruption issues.

Are you using any non-standard Apache modules?

**Comment by tomhughes** [comment]
Everything (apart from passenger obviously) seems to come from the main apache2 package. The full list is:

* /etc/apache2/mods-enabled/access_compat.load
* /etc/apache2/mods-enabled/alias.load
* /etc/apache2/mods-enabled/auth_basic.load
* /etc/apache2/mods-enabled/authn_core.load
* /etc/apache2/mods-enabled/authn_file.load
* /etc/apache2/mods-enabled/authz_core.load
* /etc/apache2/mods-enabled/authz_groupfile.load
* /etc/apache2/mods-enabled/authz_host.load
* /etc/apache2/mods-enabled/authz_user.load
* /etc/apache2/mods-enabled/autoindex.load
* /etc/apache2/mods-enabled/cgid.load
* /etc/apache2/mods-enabled/deflate.load
* /etc/apache2/mods-enabled/dir.load
* /etc/apache2/mods-enabled/env.load
* /etc/apache2/mods-enabled/expires.load
* /etc/apache2/mods-enabled/filter.load
* /etc/apache2/mods-enabled/headers.load
* /etc/apache2/mods-enabled/info.load
* /etc/apache2/mods-enabled/lbmethod_bybusyness.load
* /etc/apache2/mods-enabled/lbmethod_byrequests.load
* /etc/apache2/mods-enabled/mime.load
* /etc/apache2/mods-enabled/mpm_event.load
* /etc/apache2/mods-enabled/negotiation.load
* /etc/apache2/mods-enabled/passenger.load
* /etc/apache2/mods-enabled/proxy_balancer.load
* /etc/apache2/mods-enabled/proxy_http.load
* /etc/apache2/mods-enabled/proxy.load
* /etc/apache2/mods-enabled/rewrite.load
* /etc/apache2/mods-enabled/setenvif.load
* /etc/apache2/mods-enabled/slotmem_shm.load
* /etc/apache2/mods-enabled/socache_shmcb.load
* /etc/apache2/mods-enabled/ssl.load
* /etc/apache2/mods-enabled/status.load

We did have `proxy_fcgi` as well but I just turned that off as it isn't used on those machines any more.

**Comment by FooBarWidget** [comment]
I've gotten rid of the envvar caching in order to fix issue #1446. It might have an effect on your issue as well. Could you test the `GH-1446` branch (specifically commit 8dc1da667) and check whether the issue is gone?

**Comment by tomhughes** [comment]
It looks like that has fixed it, yes. I've done a dozen or more reloads now without any problems - previously it only took two or three before I triggered it.
